tests/macros/utils.js

```diff
@@ -63,15 +65,37 @@ function createMacroTestObject(macroName) {
         url: 'https://developer.mozilla.org/'
     };
     let environment = new Environment(pageContext, templates, true);
+    let ctx = environment.prototypeEnvironment;
+
+    /** @type {Map<string, string>} */
+    const macroResults = new Map();
+    const realTemplate = ctx.template;
+    ctx.template = jest.fn(async (name, ...args) => {
+        let macroName = String(name)
+            .replace(/:/g, '-')
+            .toLowerCase();
+        return macroResults.get(macroName) || realTemplate(name, ...args);
```

Comment:
I think you should have an `await` before realTemplate. Otherwise, one case returns a string and the other case returns a Promise<string>. It may work, but would probably make a type checker angry.

Reply:
Also: do you think it will ever be important to be able to mock a template so that it returns an empty string? That won't be possible just using || here. How about if the get() call returns null or undefined then you call realTemplate(), but the empty string is a legal value?

Reply:
Since the function is `async`, returning a promise or non&#x2011;promise is the same as returning a promise in both cases or using `await` (Promises are auto&#x2011;flattened in&nbsp;JS)

Reply:
Also, I’ll handle that case too.

(I keep forgetting that `!!''` results in&nbsp;`false` in JavaScript)

Reply:
I understand that you *can* return a promise or non-promise from an async function. But I think it is bad style to use promise and non-promise on a single line like this, and I suspect that a type checker like Flow would not allow it. Please add the await.

Reply:
TypeScript is happy with it, also <code>return&nbsp;await &lt;promise&#x2011;result&gt;</code> [is&nbsp;generally not&nbsp;recommended](https://eslint.org/docs/rules/no-return-await).

Also, TypeScript is&nbsp;fine with&nbsp;it.

Reply:
I'm not sure I believe that there is a real performance issue with `return await`, but it seems reasonable enough to avoid it. I'm fine if you want to do an if/else and return a string from one branch and a promise from the other.  My objection is returning and || expression whose value is either a string or a promise. That seems like bad style and something that will cause confusion.